### ISSUE TITLE
FR: add retry and timeout options to the tailscale_device data source

### ISSUE THREAD

**Issue by rsyring** [davidsbond/terraform-provider-tailscale#69]
**Is your feature request related to a problem? Please describe.**

My end goal is to be able to define EC2 instances in Terraform and setup DNS entries for the tailscale IP address.  The current problem I have is that, unlike an AWS public or private IP, the tailscale IP address is unknown to AWS and not returned by their APIs.  Furthermore, there is a delay between the time the instance is created and it's user_data script can run  `tailscale up ...`.  So, using the `tailscale_device` data source in the same script usually results in that API call failing.

**Describe the solution you'd like**
I'd like to be able to do something like:

```hcl
resource "aws_instance" "enterprise" {
    # ...snip...
    user_data = templatefile("tailscale-init.sh", {
        hostname = "enterprise"
        tailscale_auth_key = '...snip...'
    })
}


resource "aws_route53_record" "enterprise_pub" {
    zone_id = var.some_zone_id
    name = "enterprise"
    type = "A"
    ttl = "300"
    records = [
        aws_instance.enterprise.public_ip
    ]
}


data "tailscale_device" "enterprise" {
  name = "enterprise.${var.some_zone_domain}"
  # Retry API call 10 times waiting 10 seconds in-between each try
  retry_count = 10
  retry_interval = "10s"
}


resource "aws_route53_record" "enterprise_ts" {
    zone_id = var.some_zone_id
    name = "enterprise.ts"
    type = "A"
    ttl = "300"
    records = [
        # Another FR: should have .addresses_ipv4?
        tailscale_device.enterprise.addresses[0]
    ]
}
```

**Additional context**

If the above doesn't seem like the right approach, I'd be open to other ideas about how to provision an EC2 instance and get DNS setup for the Tailscale IP.

**Comment by davidsbond** [comment]
Hey @rsyring, I don't think these fields should be present in the `tailscale_device` data item. It seems like an appropriate use of the [depends_on](https://www.terraform.io/language/meta-arguments/depends_on) meta argument.

I could potentially add such configuration to the `provider` block, so it could be used in the provider configuration itself:

```terraform
terraform {
  required_providers {
    tailscale = {
      source = "davidsbond/tailscale"
      version = "0.7.0"
    }
  }
}

provider "tailscale" {
  retry_count = 10
  retry_interval = "10s"
}
```

You could also use a [remote-exec provisioner](https://www.terraform.io/language/resources/provisioners/remote-exec) that ensures the Tailscale device is up and running in the AWS instance.

**Comment by rsyring** [comment]
Hi David, thanks for all of your replies.  Since this one seems to be the crux of a fully managed solution using Terraform, I figured I'd start my replies here.

> It seems like an appropriate use of the [depends_on](https://www.terraform.io/language/meta-arguments/depends_on) meta argument.

I agree that `depends_on` seems like an appropriate use here.  The problem I'm currently running into is that the `aws_instance` isn't aware of or concerned with Tailscale.  So it's going to return when the instance is created and that's almost always going to be prior to tailscale being ready on the new instance.

> The remote-exec provisioner requires a [connection](https://www.terraform.io/language/resources/provisioners/connection) and supports both ssh and winrm

There is a chicken/egg problem here though.  My goal is to be able to setup servers that have no public access mechanism...their only access would be through Tailscale.  The remote-exec feature requires an SSH session.  I don't want to have to setup a VPN to get SSH access, I already have one of those and want to replace it with Tailscale.  :)

The order of operations needs to be:

- Server creation
- Run tailscale (so IP can get assigned)
- Figure out the IP of the device that just came online (but can't do this earlier than running tailscale...but, don't know when Tailscale has finished on the report server)
- Assign that IP to a DNS record

I only see two solutions, but vantage point may be limited:

* Create a way in the Tailscale API to pre-provision a device which would result in the IP assigment.  Return some kind of unique device identifier and then pass that identifier through to the new server instance and the tailscale install.  My guess is this would require a lot of upstream changes.
* Make `tailscale_device` aware of whether the device exists (at which point a TS IP address for it also exists).  If not, have it retry the API periodically until the device appears (with a timeout of some kind of course).
  * In this case, it probably still makes sense to use an implicit or explicit `depends_on` to avoid polling the TS API before the device has been created.

Thanks for the discussion.

**Comment by davidsbond** [comment]
Hey @rsyring, thanks for the detailed reply. I agree with you that this is one of those typical terraform problems where you rely on an asynchronous thing happening after the creation of a resource. I can also see it being a common use-case and I agree it defeats the point if you have to initially expose the AWS instance. 

> Create a way in the Tailscale API to pre-provision a device which would result in the IP assigment. Return some kind of unique device identifier and then pass that identifier through to the new server instance and the tailscale install. My guess is this would require a lot of upstream changes.

Unfortunately as I don't work for Tailscale, I can't implement such an API, but it might be worth opening an issue on their repository and starting a discussion there.

> Make tailscale_device aware of whether the device exists (at which point a TS IP address for it also exists). If not, have it retry the API periodically until the device appears (with a timeout of some kind of course).
In this case, it probably still makes sense to use an implicit or explicit depends_on to avoid polling the TS API before the device has been created.

I think this is closer to the solution you will need, but I don't think that it should necessarily live in the `tailscale_device` resource. To me having configuration within individual resources/data resources for things like retries and backoff doesn't feel like the "terraform" way to do it. But I am also not sure what the best solution would be at this time.

Since the Tailscale API itself can't support you, it might be for the best to provide that level of configuration for API calls in the `tailscale_device` data resource. If the Tailscale API then supports you better in future, we could deprecate those.

To summarise, I can start with an initial implementation that looks like this:

```terraform
data "tailscale_device" "example" {
  name = "my-device"
  retry_for = "60s"
}
```

This would mean that the Tailscale API is polled regularly, but if the device cannot be found after 60 seconds the `apply` command would fail. I don't think I want to add configuration for number of attempts initially, I'd rather keep the configuration surface small.

Would you be happy with this as a first attempt?

**Comment by rsyring** [comment]
> Would you be happy with this as a first attempt?

Yes, seems like a good place to start.  Thank you.

**Comment by rsyring** [comment]
> I don't think I want to add configuration for number of attempts initially,

You know what, I didn't mean to have a count.  I meant to have a total timeout and a polling interval.

```hcl
data "tailscale_device" "enterprise" {
  name = "enterprise.${var.some_zone_domain}"
  # Retry API call 10 times waiting 10 seconds in-between each try
  retry_for = "60s"
  retry_every= "10s"
}
``` 

I suppose you could hard code the "every" but I envision that being something would ask for control over.  As long as it's a reasonably short interval, I don't mind it being hard coded.  I'd like any way to be able to bring this all together!  :)

Thanks.
